server/src/main/java/org/apache/iotdb/db/auth/AuthorizerManager.java

```diff
@@ -215,131 +245,312 @@ public void reset() throws AuthException {
     try {
       iAuthorizer.processLoadSnapshot(snapshotDir);
     } finally {
-      snapshotLock.writeLock().unlock();
+      authReadWriteLock.writeLock().unlock();
+    }
+  }
+
+  public TSStatus checkPath(String username, List<String> allPath, int permission)
+      throws AuthException {
+    authReadWriteLock.readLock().lock();
+    try {
+      User user = userCache.getIfPresent(username);
+      boolean status = true;
+      if (user != null) {
+        for (String path : allPath) {
+          if (!user.checkPrivilege(path, permission)) {
+            for (String roleName : user.getRoleList()) {
+              Role role = roleCache.getIfPresent(roleName);
+              // It is detected that the role of the user does not exist in the cache, indicating
+              // that the permission information of the role has changed.
+              // The user cache needs to be initialized
+              if (role == null) {
+                invalidateCache(username, "");
+                status = false;
+                break;
+              }
+              if (role.checkPrivilege(path, permission)) {
+                status = true;
+                break;
+              } else {
+                status = false;
+              }
+            }
+            // Prove that neither the user nor his role has permission
+            if (!status) {
+              break;
+            }
+          }
+        }
+      }
+      if (status) {
+        return RpcUtils.getStatus(TSStatusCode.SUCCESS_STATUS);
+      }
+      tPermissionInfoResp = ClusterAuthorizer.checkPath(username, allPath, permission);
+      if (tPermissionInfoResp.getStatus().getCode()
+          == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+        userCache.get(username);
+        return tPermissionInfoResp.getStatus();
+      } else {
+        return tPermissionInfoResp.getStatus();
+      }
+    } finally {
+      authReadWriteLock.readLock().unlock();
+    }
+  }
+
+  /** Check the user */
+  public TSStatus checkUser(String username, String password) {
+    authReadWriteLock.readLock().lock();
+    try {
+      User user = userCache.getIfPresent(username);
+      if (user != null
+          && password != null
+          && AuthUtils.validatePassword(password, user.getPassword())) {
+        return RpcUtils.getStatus(TSStatusCode.SUCCESS_STATUS);
+      }
+      tPermissionInfoResp = ClusterAuthorizer.checkUser(username, password);
+      if (tPermissionInfoResp.getStatus().getCode()
+          == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+        userCache.get(username);
+        return tPermissionInfoResp.getStatus();
+      } else {
+        return tPermissionInfoResp.getStatus();
+      }
+    } finally {
+      authReadWriteLock.readLock().unlock();
+    }
+  }
+
+  public SettableFuture<ConfigTaskResult> queryPermission(
+      TAuthorizerReq authorizerReq, ConfigNodeClient configNodeClient) {
+    authReadWriteLock.writeLock().lock();
```

Comment:
Maybe the readLock() should be used here ?

Reply:
OK